unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -211,4 +215,57 @@ public boolean equals(final Object other) {
   public int hashCode() {
     return Arrays.hashCode(bytes);
   }
+
+  private char[] getSoundexMapping() {
+    return this.soundexMapping;
+  }
+
+  private char map(final char ch) {
+    final int index = ch - 'A';
+    if (index < 0 || index >= this.getSoundexMapping().length) {
+      throw new IllegalArgumentException("The character is not mapped: " + ch);
+    }
+    return this.getSoundexMapping()[index];
+  }
+
+  private char getMappingCode(final String str, final int index) {
+    // map() throws IllegalArgumentException
+    final char mappedChar = this.map(str.charAt(index));
+    // HW rule check
+    if (index > 1 && mappedChar != '0') {
+      final char hwChar = str.charAt(index - 1);
+      if ('H' == hwChar || 'W' == hwChar) {
+        final char preHWChar = str.charAt(index - 2);
+        final char firstCode = this.map(preHWChar);
+        if (firstCode == mappedChar || 'H' == preHWChar || 'W' == preHWChar) {
+          return 0;
+        }
+      }
+    }
+    return mappedChar;
+  }
+
+  public String soundex(String str) {
```

Comment:
We don't need to pass in the `str`, what we need to handle is the `UTF8String` itself.